src/core/common/message.cpp

```diff
@@ -254,23 +269,9 @@ otError Message::SetLength(uint16_t aLength)
 {
     otError  error              = OT_ERROR_NONE;
     uint16_t totalLengthRequest = GetReserved() + aLength;
-    uint16_t totalLengthCurrent = GetReserved() + GetLength();
-    int      bufs               = 0;
 
     VerifyOrExit(totalLengthRequest >= GetReserved(), error = OT_ERROR_INVALID_ARGS);
 
-    if (totalLengthRequest > kHeadBufferDataSize)
-    {
-        bufs = (((totalLengthRequest - kHeadBufferDataSize) - 1) / kBufferDataSize) + 1;
-    }
-
-    if (totalLengthCurrent > kHeadBufferDataSize)
-    {
-        bufs -= (((totalLengthCurrent - kHeadBufferDataSize) - 1) / kBufferDataSize) + 1;
-    }
-
-    SuccessOrExit(error = GetMessagePool()->ReclaimBuffers(bufs, GetPriority()));
```

Comment:
why is it not necessary anymore?

Reply:
Pre-reclaim here does not make much difference from per buffer reclaim when required in `ResizeMessage()`. @simonlingoogle 

Just for discussion:
- One possible optimization is to make this piece of code and `MessagePool::ReclaimBuffers(int aNumBuffers, Message::Priority aPriority)` more meaningful is to add one method to get the available buffers evict-able, and abort earlier if (the available evict-able buffers + existing free buffers) could not meet the buffer requirements, thus avoid meaningless eviction.
- Is it necessary to keep `MessagePool::ReclaimBuffers(int aNumBuffers, Message::Priority aPriority)` if given existing per buffer reclaim/allocation implementation?

Thoughts? @bukepo @simonlingoogle

Reply:
When reclaimation is required, pre-reclaimation here doesn't make much difference from per buffer reclaimation in `ResizeMessage()`

Reply:
I don't think the current heap allocator can give a precise number of buffers it can allocate.

Reply:
To make the pre-check here more useful is to introduce some method to check whether or not evict-able buffers could meet the requirement before eviction if the existing available free buffers could not meet the requirement I think, thus avoiding meaningless eviction if it is predictable it could not meet the buffer requirement.

Reply:
> I don't think the current heap allocator can give a precise number of buffers it can allocate.

It should be possible to enhance the allocator to do that?

@librasungirl Predicting if if eviction should be possible. If we can get the number of message buffers that the current heap can allocate (N), then the max number of message buffers we can get after eviction is `N + Number of message buffers of lower priority`.

Reply:
IMO, I think this is fine as is, but I like @librasungirl  suggestion above (possibly as a future enhancement).

Reply:
But should we keep the original code? @abtink @bukepo @librasungirl

Reply:
ResizeMessage() should be simplified if we make sure memory is enough before allocating. The code is well kept in history, so I tend not to keep it just for future reference.

Reply:
Even with the heap allocator, if `ReclaimBuffers` failed, we **can be sure** that we are not going to be able to `ResizeMessage`. 

Furthermore, evicting messages of lower priority can still increase the chance of having enough buffer for a higher-priority message. The the code is still helpful even for `MESSAGE_USE_HEAP=ON`

Not to mention, `MESSAGE_USE_HEAP=ON` is not the only implementation. When `MESSAGE_USE_HEAP=OFF`, the original code works just fine. Why remove the code if it's helpful in both cases? 

Thoughts?

Reply:
>  if ReclaimBuffers failed, we can be sure that we are not going to be able to ResizeMessage.

Why requiring knowing `ReclaimBuffers failed` in advance before `ResizeMessage`? 

There is `ReclaimBuffer` when requesting new buffer in ResizeMessage() if required, and NO_BUFS error would be returned if the requirement could not be met.